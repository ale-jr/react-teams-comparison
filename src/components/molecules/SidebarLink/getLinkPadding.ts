import { Theme } from "interfaces/theme";

interface GetLinkPaddingAtributes {
  theme: Theme;
  level: number;
  $hasIcon: boolean;
}
export const getLinkPadding = ({
  theme,
  level,
  $hasIcon,
}: GetLinkPaddingAtributes): string => {
  const calcWithIcon = `calc(${theme.spacing.md} * ${level + 1})`;
  const calcWithoutIcon = `calc(1.25em + ${calcWithIcon})`;
  return $hasIcon ? calcWithIcon : calcWithoutIcon;
};
