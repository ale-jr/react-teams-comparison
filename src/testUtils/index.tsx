import { render, RenderOptions, RenderResult } from '@testing-library/react'
import ThemeProvider from 'providers/ThemeProvider'
import { ReactElement } from 'react'


const Providers: React.FC<{
    children: JSX.Element | JSX.Element[]
}> = ({ children }) => <ThemeProvider>{children}</ThemeProvider>


const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "queries">): RenderResult => render(ui, {
    wrapper: Providers,
    ...options
})

export * from '@testing-library/react'
export {customRender as render}