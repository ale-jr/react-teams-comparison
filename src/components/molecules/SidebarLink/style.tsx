import styled, { css, keyframes } from "styled-components";
import { Link as ReactRouterLink } from 'react-router-dom'
import { getLinkPadding } from './getLinkPadding'


export const LinkContainer = styled.li`
    list-style-type: none;
`

const slideIn = keyframes`
  0% {
    transform: translateX(-4px);
    opacity: 0;
  }
  100% {
    transform: translateX(0);
    opacity: 1;
  }
`
const slideOut = keyframes`
  0% {
    transform: translateX(0);
    opacity: 1;
  }

 100% {
    transform: translateX(-4px);
    opacity: 0;
  }
`


interface LinkProps {
  level: number;
  $hasIcon: boolean;
  $active: boolean;
}
export const Link = styled(ReactRouterLink) <LinkProps>`
    //Position and Layout
    position: relative;

    //Display and visibility
    display: inline-flex;
    align-items: center;

    //Clipping


    //Animation
    transition: all ${({ theme }) => theme.animation.shorthand};

    //Box model (From outside in)
    padding: ${({ theme }) => `${theme.spacing.sm} ${theme.spacing.xl}`};
    padding-left: ${getLinkPadding};


    //Background


    //Typography
    color: ${({ $active, theme }) => $active ? theme.colors.neutral["600"] : theme.colors.neutral["500"]};
    font-weight: ${({ theme, $active }) => $active ? theme.typography.weight.bold : theme.typography.weight.regular};
    font-size: ${({ theme }) => theme.typography.size.md};
    text-decoration: none;
    white-space: nowrap;


    //Pseudo-classes and pseudo-elements
    &:hover{
        color: ${({ theme }) => theme.colors.primary.default};
    }

    span.title{
        padding-left: ${({ theme }) => theme.spacing.md};
    }

    svg.icon{
        font-size: ${({ theme }) => theme.typography.size.xl};
    }

    &::before{
            animation: ${({ $active }) => css`${$active ? slideIn : slideOut} 125ms linear both`};
            content: "";
            position: absolute;
            left: 0;
            top: 0;
            height: 100%;
            width: 4px;
            background-color: ${({ theme }) => theme.colors.primary.default};
        }
`
