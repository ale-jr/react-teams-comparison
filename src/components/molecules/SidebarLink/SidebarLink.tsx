import { IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { LinkContainer, Link } from './style'

interface SidebarLinkProps {
    to: string
    icon?: IconDefinition,
    title: string,
    active?: boolean,
    level?: number
}

const SidebarLink: React.FC<SidebarLinkProps> = ({ icon, title, to, active = false, level = 0 }) => {
    return <LinkContainer>
        <Link to={to} level={level} $hasIcon={!!icon} $active={active}>
            {icon && <FontAwesomeIcon icon={icon} className="icon" fixedWidth />}
            <span className="title">{title}</span>
        </Link>
    </LinkContainer>
}


export default SidebarLink