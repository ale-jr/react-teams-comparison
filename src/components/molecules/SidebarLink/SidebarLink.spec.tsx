import 'jest-styled-components'
import { render } from "testUtils"
import SidebarLink from './index'
import { getLinkPadding } from "./getLinkPadding"
import defaultTheme from 'themes/default'

import { faHome } from "@fortawesome/free-solid-svg-icons"

describe("SidebarLink", () => {

    it("Should match snapshot", () => {
        const { asFragment } = render(<SidebarLink title='Home' to='/home' icon={faHome} />)
        expect(asFragment()).toMatchSnapshot()
    })

    it("Should match snapshot (active link)", () => {
        const { asFragment } = render(<SidebarLink title='Home' to='/home' icon={faHome} active />)
        expect(asFragment()).toMatchSnapshot()
    })

    it("Should match snapshot (sublink)", () => {
        const { asFragment } = render(<SidebarLink title='Home' to='/home' icon={faHome} active level={1} />)
        expect(asFragment()).toMatchSnapshot()
    })

    it("Should generate padding calc correctly", () => {
        const assertions = [
            {
                level: 0,
                $hasIcon: true,
                assertion: 'calc(1em * 1)'
            },
            {
                level: 1,
                $hasIcon: true,
                assertion: 'calc(1em * 2)'
            },
            {
                level: 0,
                $hasIcon: false,
                assertion: 'calc(1.25em + calc(1em * 1))'
            },
            {
                level: 1,
                $hasIcon: false,
                assertion: 'calc(1.25em + calc(1em * 2))'
            },
        ]

        assertions.forEach(({ level, $hasIcon, assertion }) => {
            const result = getLinkPadding({ level, $hasIcon, theme: defaultTheme })
            expect(result).toBe(assertion)
        })
    })
})