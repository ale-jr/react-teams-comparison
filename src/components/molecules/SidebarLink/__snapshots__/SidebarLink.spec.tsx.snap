// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarLink Should match snapshot (active link) 1`] = `
<DocumentFragment>
  .c0 {
  list-style-type: none;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 120ms linear;
  transition: all 120ms linear;
  padding: 0.5em 2em;
  padding-left: calc(1em * 1);
  color: #1C1E2A;
  font-weight: 700;
  font-size: 1em;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c1:hover {
  color: #2678FF;
}

.c1 span.title {
  padding-left: 1em;
}

.c1 svg.icon {
  font-size: 1.5em;
}

.c1::before {
  -webkit-animation: hTmvpY 125ms linear both;
  animation: hTmvpY 125ms linear both;
  content: "";
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 4px;
  background-color: #2678FF;
}

<li
    class="c0"
  >
    <a
      class="c1"
      href="/home"
      level="0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-house fa-fw icon"
        data-icon="house"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M575.8 255.5C575.8 273.5 560.8 287.6 543.8 287.6H511.8L512.5 447.7C512.5 450.5 512.3 453.1 512 455.8V472C512 494.1 494.1 512 472 512H456C454.9 512 453.8 511.1 452.7 511.9C451.3 511.1 449.9 512 448.5 512H392C369.9 512 352 494.1 352 472V384C352 366.3 337.7 352 320 352H256C238.3 352 224 366.3 224 384V472C224 494.1 206.1 512 184 512H128.1C126.6 512 125.1 511.9 123.6 511.8C122.4 511.9 121.2 512 120 512H104C81.91 512 64 494.1 64 472V360C64 359.1 64.03 358.1 64.09 357.2V287.6H32.05C14.02 287.6 0 273.5 0 255.5C0 246.5 3.004 238.5 10.01 231.5L266.4 8.016C273.4 1.002 281.4 0 288.4 0C295.4 0 303.4 2.004 309.5 7.014L564.8 231.5C572.8 238.5 576.9 246.5 575.8 255.5L575.8 255.5z"
          fill="currentColor"
        />
      </svg>
      <span
        class="title"
      >
        Home
      </span>
    </a>
  </li>
</DocumentFragment>
`;

exports[`SidebarLink Should match snapshot (sublink) 1`] = `
<DocumentFragment>
  .c0 {
  list-style-type: none;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 120ms linear;
  transition: all 120ms linear;
  padding: 0.5em 2em;
  padding-left: calc(1em * 2);
  color: #1C1E2A;
  font-weight: 700;
  font-size: 1em;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c1:hover {
  color: #2678FF;
}

.c1 span.title {
  padding-left: 1em;
}

.c1 svg.icon {
  font-size: 1.5em;
}

.c1::before {
  -webkit-animation: hTmvpY 125ms linear both;
  animation: hTmvpY 125ms linear both;
  content: "";
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 4px;
  background-color: #2678FF;
}

<li
    class="c0"
  >
    <a
      class="c1"
      href="/home"
      level="1"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-house fa-fw icon"
        data-icon="house"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M575.8 255.5C575.8 273.5 560.8 287.6 543.8 287.6H511.8L512.5 447.7C512.5 450.5 512.3 453.1 512 455.8V472C512 494.1 494.1 512 472 512H456C454.9 512 453.8 511.1 452.7 511.9C451.3 511.1 449.9 512 448.5 512H392C369.9 512 352 494.1 352 472V384C352 366.3 337.7 352 320 352H256C238.3 352 224 366.3 224 384V472C224 494.1 206.1 512 184 512H128.1C126.6 512 125.1 511.9 123.6 511.8C122.4 511.9 121.2 512 120 512H104C81.91 512 64 494.1 64 472V360C64 359.1 64.03 358.1 64.09 357.2V287.6H32.05C14.02 287.6 0 273.5 0 255.5C0 246.5 3.004 238.5 10.01 231.5L266.4 8.016C273.4 1.002 281.4 0 288.4 0C295.4 0 303.4 2.004 309.5 7.014L564.8 231.5C572.8 238.5 576.9 246.5 575.8 255.5L575.8 255.5z"
          fill="currentColor"
        />
      </svg>
      <span
        class="title"
      >
        Home
      </span>
    </a>
  </li>
</DocumentFragment>
`;

exports[`SidebarLink Should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  list-style-type: none;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 120ms linear;
  transition: all 120ms linear;
  padding: 0.5em 2em;
  padding-left: calc(1em * 1);
  color: #62656C;
  font-weight: 400;
  font-size: 1em;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c1:hover {
  color: #2678FF;
}

.c1 span.title {
  padding-left: 1em;
}

.c1 svg.icon {
  font-size: 1.5em;
}

.c1::before {
  -webkit-animation: dcGRDW 125ms linear both;
  animation: dcGRDW 125ms linear both;
  content: "";
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 4px;
  background-color: #2678FF;
}

<li
    class="c0"
  >
    <a
      class="c1"
      href="/home"
      level="0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-house fa-fw icon"
        data-icon="house"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M575.8 255.5C575.8 273.5 560.8 287.6 543.8 287.6H511.8L512.5 447.7C512.5 450.5 512.3 453.1 512 455.8V472C512 494.1 494.1 512 472 512H456C454.9 512 453.8 511.1 452.7 511.9C451.3 511.1 449.9 512 448.5 512H392C369.9 512 352 494.1 352 472V384C352 366.3 337.7 352 320 352H256C238.3 352 224 366.3 224 384V472C224 494.1 206.1 512 184 512H128.1C126.6 512 125.1 511.9 123.6 511.8C122.4 511.9 121.2 512 120 512H104C81.91 512 64 494.1 64 472V360C64 359.1 64.03 358.1 64.09 357.2V287.6H32.05C14.02 287.6 0 273.5 0 255.5C0 246.5 3.004 238.5 10.01 231.5L266.4 8.016C273.4 1.002 281.4 0 288.4 0C295.4 0 303.4 2.004 309.5 7.014L564.8 231.5C572.8 238.5 576.9 246.5 575.8 255.5L575.8 255.5z"
          fill="currentColor"
        />
      </svg>
      <span
        class="title"
      >
        Home
      </span>
    </a>
  </li>
</DocumentFragment>
`;
